velox/exec/MergingVectorOutput.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/MergingVectorOutput.h"
+
+namespace facebook::velox::exec {
+
+MergingVectorOutput::MergingVectorOutput(
+    velox::memory::MemoryPool* pool,
+    int64_t preferredOutputBatchBytes,
+    int32_t preferredOutputBatchRows,
+    int64_t minOutputBatchBytes,
+    int32_t minOutputBatchRows)
+    : pool_(pool),
+      preferredOutputBatchBytes_(preferredOutputBatchBytes),
+      preferredOutputBatchRows_(preferredOutputBatchRows),
+      minOutputBatchBytes_(minOutputBatchBytes),
+      minOutputBatchRows_(minOutputBatchRows) {}
+
+RowVectorPtr MergingVectorOutput::getOutput() {
+  if (noMoreInput_) {
+    flush();
+  }
+  if (outputQueue_.size() > 0) {
+    const auto res = outputQueue_.front();
+    outputQueue_.pop();
+    return res;
+  }
+  return nullptr;
+}
+
+void MergingVectorOutput::addVector(RowVectorPtr vector) {
+  const auto numInput = vector->size();
+  if (numInput == 0) {
+    return;
+  }
+
+  // Avoid memory copying for pages that are big enough
+  if (numInput >= minOutputBatchRows_ ||
+      vector->estimateFlatSize() >= minOutputBatchBytes_) {
+    outputQueue_.push(std::move(vector));
+    return;
+  }
+
+  buffer(std::move(vector));
+}
+
+void MergingVectorOutput::buffer(RowVectorPtr vector) {
+  const auto numInput = vector->size();
+  const auto& children = vector->children();
+
+  // In order to prevent vector.resize() cause memory copy, the initial value
+  // of bufferInputs_ row count is preferredOutputBatchRows_ and cannot exceed
+  // preferredOutputBatchRows_.
+  if (bufferRows_ + numInput > preferredOutputBatchRows_) {
+    flush();
+  }
+
+  if (!bufferInputs_) {
+    bufferInputs_ = BaseVector::create<RowVector>(
+        vector->type(), preferredOutputBatchRows_, pool_);
+  }
+
+  bufferInputs_->resize(bufferRows_ + numInput);
```

Comment:
Extract the merging logic to a separate function that can be reused.

Also check encoding of each column, if any of them has non-flat encoding, do not merge them

Reply:
The output vector of filtering operators is generally DictionaryVector. If DictionaryVector is not merged, there will be no optimization.

Reply:
Then you should measure whether it is really beneficial to merge the dictionary vectors.  My hunch is that it is only beneficial when the selectivity is lower than a threshold, however it is very hard to hard-code a value not relying on adaptivity.

Reply:
@Yuhta By 'measure whether it is really beneficial to merge the dictionary vectors', you mean measure the downstream operator's performance at runtime right? My understanding is comparing the performance of an operator before and after turn on a config for some duration at runtime. Do we have this in Velox now or we have implemented it in this or another PR?

Reply:
The adaptivity mechanism for this case is not there and we need to think about it.  For now I would suggest to leave `minMergingVectorOutputBatchRows` or similar as mutable state that can be changed from driver, because the measurement probably needs to be done at driver level.

The vector merging treatment here is somehow separate, there are some encodings we probably do not want to merge (lazy), and some others we want to merge conditionally depending on the data distribution (constant/dictionary).  The better we do the merging here, the adaptivity will favor more on the merging path.

Reply:
done

Reply:
done